packages/js/src/utils/find-npm-dependencies.ts

```diff
@@ -0,0 +1,203 @@
+        if (projectDep) collectAll(projectDep, collectedDeps);
+      }
+    }
+  }
+
+  collectAll(sourceProject, results);
+
+  return results;
+}
+
+// Keep track of workspace libs we already read package.json for so we don't read from disk again.
+const seenWorkspaceDeps: Record<string, { name: string; version: string }> = {};
+
+// ts, js, mts, mjs, cts, cjs, tsx, jsx
+const sourceFileRegex = /\.([jt]sx?|[ct][jt]s)$/;
+
+function collectDependenciesFromFileMap(
+  workspaceRoot: string,
+  sourceProject: ProjectGraphProjectNode,
+  projectGraph: ProjectGraph,
+  projectFileMap: ProjectFileMap,
+  ignoreFileFilters: ((file: string) => boolean)[],
+  npmDeps: Record<string, string>
+): void {
+  const files = projectFileMap[sourceProject.name];
+  if (!files) return;
+
+  for (const fileData of files) {
+    if (!fileData.deps) continue;
+    if (!sourceFileRegex.test(fileData.file)) continue;
+    if (ignoreFileFilters.some((f) => f(fileData.file))) continue;
+
+    for (const dep of fileData.deps) {
+      const target = fileDataDepTarget(dep);
+
+      // If the node is external, then read package info from `data`.
+      const externalDep = projectGraph.externalNodes[target];
+      if (externalDep?.type === 'npm') {
+        npmDeps[externalDep.data.packageName] = externalDep.data.version;
+        continue;
+      }
+
+      // If node is internal, then try reading package info from `package.json` (which must exist for this to work).
+      const workspaceDep = projectGraph.nodes[target];
+      if (!workspaceDep) continue;
+      const cached = seenWorkspaceDeps[workspaceDep.name];
+      if (cached) {
+        npmDeps[cached.name] = cached.version;
+      } else {
+        const packageJson = readPackageJson(workspaceDep, workspaceRoot);
+        if (packageJson) {
+          // This is a workspace lib so we can't reliably read in a specific version since it depends on how the workspace is set up.
+          // ASSUMPTION: Most users will use '*' for workspace lib versions. Otherwise, they can manually update it.
+          npmDeps[packageJson.name] = '*';
+          seenWorkspaceDeps[workspaceDep.name] = {
+            name: packageJson.name,
+            version: '*',
+          };
+        }
+      }
+    }
+  }
+}
+
+function readPackageJson(
+  project: ProjectGraphProjectNode,
+  workspaceRoot: string
+): null | {
+  name: string;
+  dependencies?: Record<string, string>;
+  optionalDependencies?: Record<string, string>;
+  peerDependencies?: Record<string, string>;
+} {
+  const packageJsonPath = join(
+    workspaceRoot,
+    project.data.root,
+    'package.json'
+  );
+  if (fileExists(packageJsonPath)) return readJsonFile(packageJsonPath);
+  return null;
+}
+
+function collectPeerAndOptionalDependencies(
+  workspaceRoot: string,
+  sourceProject: ProjectGraphProjectNode,
+  projectGraph: ProjectGraph,
+  npmDeps: Record<string, string>
+): void {
+  const packageJson = readPackageJson(sourceProject, workspaceRoot);
```

Comment:
What would be the purpose of the function?

Whatever I add manually as a `peer` or `optional` dependency in `package.json` will be propagated back to the linter rule as a "detected" dependency?

Reply:
I was getting a lot of optional dependencies and peer dependencies being incorrectly picked up as missing from `package.json`.

e.g.

```
> nx lint nx
  34:3  error  The "nx" uses the following packages, but they are missing from the "dependencies":
    - @nx/nx-win32-x64-msvc
    - @nx/nx-win32-arm64-msvc
    - @nx/nx-darwin-x64
    - @nx/nx-darwin-arm64
    - @nx/nx-freebsd-x64
    - @nx/nx-linux-x64-musl
    - @nx/nx-linux-x64-gnu
    - @nx/nx-linux-arm64-musl
    - @nx/nx-linux-arm64-gnu
    - @nx/nx-linux-arm-gnueabihf  @nx/dependency-checks
```

It could be another problem though so I can take a closer look.

Reply:
Okay, it was due to `getAllDependencies` not also picking up `optionalDependencies`. Updated that function instead.